tests/rust.rs

```diff
@@ -0,0 +1,85 @@
+use anyhow::Context;
+
+use std::path::PathBuf;
+
+struct RustTestCase {
+    name: String,
+    humble_spec: PathBuf,
+    humble_rust_out: PathBuf,
+    main: PathBuf,
+}
+
+impl RustTestCase {
+    fn run(&self) {
+        let spec_file = std::fs::read_to_string(&self.humble_spec).expect("read humble spec");
+        let spec = humblegen::parse_spec_str_or_panic(&spec_file);
+        let spec_rust = humblegen::Language::Rust.render(&spec).to_string();
+        // TODO rustfmt
+        std::fs::write(&self.humble_rust_out, spec_rust).expect("write generated rust code");
+
+        let t = trybuild::TestCases::new();
+        t.pass(&self.main);
+        // cases run on drop of t
+        drop(t);
+    }
+
+    fn from_test_dir(dir: &std::fs::DirEntry) -> anyhow::Result<RustTestCase> {
+        let entries: Vec<std::fs::DirEntry> = std::fs::read_dir(dir.path())?
+            .collect::<std::io::Result<Vec<std::fs::DirEntry>>>()
+            .unwrap()
+            .into_iter()
+            .filter(|e| e.file_type().expect("get file type").is_file())
+            .collect();
+
+        let name = dir
+            .file_name()
+            .into_string()
+            .ok()
+            .context("test case dir name must be a Rust String")?;
+
+        let mut humble_spec = None;
+        let mut humble_rust_out = None;
+        let mut main = None;
+
+        for entry in entries {
+            let name = entry
+                .file_name()
+                .into_string()
+                .ok()
+                .context("test case file names must be Rust strs")?;
+
+            match name.as_str() {
```

Comment:
This is nontrivial logic that I would like to see documented on the `struct` docs or even here, in case I want to write a test case myself some day.

Reply:
I think it's pretty clear for anyone who's looking at a git checkout instead of the GitHub review UI
![image](https://user-images.githubusercontent.com/956573/81980156-64154900-962e-11ea-8b30-1434259cfc20.png)